File: Baekjoon/분할정복/1629번_곱셈.c
#include <stdio.h>

#define true 1
#define false 0

int main() {
  int a, b, c;   scanf("%d %d %d", &a, &b, &c);
  int power = a, count = 0;
  int save[b];

  while (true) {
    power = (power * a) % c;
    save[count] = power;

    printf("%d\n", power);

    if (power == a % c || power == (a * a) % c) {
      printf("%d\n", save[b % count]);
      break;
    }

    count++;
  }

  return 0;
}


/*
#include <stdio.h>

unsigned int power(unsigned int a, unsigned int b, unsigned int c) {
  if (b == 1)
    return a % c;
  else if (b == 0)
    return 1;
  unsigned int result = 1;
  if (b % 2 == 1)
    result *= (a % c);
  result = (result * (power(a, b / 2, c) * power(a, b / 2, c))) % c;
  return result;
}

int main() {
  unsigned int a, b, c, result;  scanf("%u %u %u", &a, &b, &c);
  result = power(a, b, c);
  printf("%u\n", result);
  return 0;
}
*/

/*
#include <stdio.h>

unsigned int *memo;

void power(unsigned int a, unsigned int b, unsigned int c) {
  if (b == 0)
    return;
  memo[b] = (a % memo[b-1]) % c;
  power(a, b-1, c);
}

int main() {
  unsigned int a, b, c;  scanf("%u %u %u", &a, &b, &c);

  memo = (unsigned int*)malloc(sizeof(unsigned int)*(b+1));
  power(a, b, c);
  printf("%u\n", memo[1]);

  return 0;
}
*/

File: Baekjoon/분할정복/1629번_곱셈.py
# a, b, c = map(int, input().split())
# power = 1
# count = 0

# save = [False]

# if a < c:
#   pass
# else:
#   while True:
#     power = (power * a) % c
#     save.append(power)

#     # print(save)

#     if count != 0 and save[1] == power:
#       # print(count)
#       print(save[b % count + 2])
#       break

#     count += 1

# a, b, c = map(int, input().split())
# save = [1]
# power = 1
# count = 0

# while a < c:
#   a *= 2
#   count += 1
#   save.append(a % c)

# pivot = count

# while True:
#   power = (power * a) % c
#   save.append(power)

#   print(save)

#   # if count != 0 and save[1] == power:
#   if count != pivot and save[pivot] == power:
#     # print(count)
#     print(save[pivot + b % (count - pivot)] % c)
#     break

#   count += 1

def pow(base, expo, modular):
  if expo == 1:
    return base % modular
  elif expo % 2 == 1:
    return (base * pow(base, expo-1, modular)) % modular
  else:
    return (pow(base, expo // 2, modular) * pow(base, expo // 2, modular)) % modular

a, b, c = map(int, input().split())
print(pow(a, b, c))
